### PR TITLE
Fix camel case type warning for types with trailing underscores

### LINKED ISSUE #54099: warning: type `__` should have a camel case name such as `CamelCase`
A type with the name `_X` doesn't get a warning, but `__` does.

### PR DESCRIPTION
**Pull request by osa1**
Fixes #54099

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@osa1 your PR title says:

> trailing underscores

but I don't see any test for that, just leading underscores. Perhaps add some more tests?

**Comment by nikomatsakis** [comment]
Otherwise r=me, I agree that this makes sense to me.

**Comment by osa1** [comment]
> but I don't see any test for that, just leading underscores. Perhaps add some more tests?

Done!

File: src/test/ui/lint/issue-54099-camel-case-underscore-types.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+#![forbid(non_camel_case_types)]
+#![allow(dead_code)]
+
+// None of the following types should generate a warning
+struct _X {}
+struct __X {}
+struct __ {}
+struct X_ {}
+struct X__ {}
+struct X___ {}
```
Comment (nikomatsakis) [inline src/test/ui/lint/issue-54099-camel-case-underscore-types.rs]: looking around, I couldn't find many tests for this lint really...

Comment (nikomatsakis) [reply src/test/ui/lint/issue-54099-camel-case-underscore-types.rs]: e.g., to see whether `X_X` would fail

**Comment by nikomatsakis** [comment]
Anyway, I'm inclined to r+ this PR -- @oli-obk, @Manishearth, do we think we need an FCP here? I guess we could. Maybe I'll nominate for lang-team to discuss on Thu.

File: src/test/ui/lint/issue-54099-camel-case-underscore-types.rs (hunk 1 above)
Comment (osa1) [reply src/test/ui/lint/issue-54099-camel-case-underscore-types.rs]: Hmm I actually don't know the answer to that myself -- that is whether `X_X` is camel case or not. Btw there are a few more tests about this in `lint-non-camel-case-types.rs`.

**Comment by Manishearth** [comment]
I think this is a bug fix and can just be landed, but feel free to FCP. I
don't have a strong opinion on this.

**Comment by nikomatsakis** [comment]
We chatted in the lang team mtg and decided "oh what the heck we'll FCP it".

**Comment by nikomatsakis** [comment]
I believe @Centril's exact quote was "I have never cared less about anything"

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose we allow any number of trailing or leading underscores because...why not.

**Comment by joshtriplett** [comment]
:man_shrugging: :+1:

**Comment by eddyb** [comment]
Do we approve now, or wait some more?

**Comment by joshtriplett** [comment]
@eddyb Go ahead.

**Pull request merged**